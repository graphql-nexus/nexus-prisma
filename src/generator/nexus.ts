import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
