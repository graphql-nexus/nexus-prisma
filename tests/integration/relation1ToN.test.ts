import endent from 'endent'
import { gql } from 'graphql-tag'
import { objectType, queryType } from 'nexus'
import { testIntegration } from '../__helpers__'

testIntegration({
  description: 'can project user-to-posts relationship',
  datasourceSchema: endent`
    model User {
      id         String    @id
      posts      Post[]
    }
    model Post {
      id        String  @id
      author    User?   @relation(fields: [authorId], references: [id])
      authorId  String
    }
  `,
  datasourceSeed(prisma) {
    return prisma.user.create({
      data: {
        id: 'user1',
        posts: {
          create: [{ id: 'post1' }, { id: 'post2' }],
        },
      },
    })
  },
  apiSchema({ User, Post }) {
    return [
      queryType({
        definition(t) {
          t.list.field('users', {
            type: 'User',
            resolve(_, __, ctx) {
              return ctx.prisma.user.findMany()
            },
          })
        },
      }),
      objectType({
        name: User.$name,
        definition(t) {
          t.field(User.id.name, User.id)
          t.field(User.posts.name, User.posts)
        },
      }),
      objectType({
        name: Post.$name,
        definition(t) {
          t.field(Post.id.name, Post.id)
        },
      }),
    ]
  },
  apiClientQuery: gql`
    query {
      users {
        id
        posts {
          id
        }
      }
    }
  `,
})
