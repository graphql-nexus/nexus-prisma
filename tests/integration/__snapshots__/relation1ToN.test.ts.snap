// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can project user-to-posts relationship: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": "user1",
        "posts": Array [
          Object {
            "id": "post1",
          },
          Object {
            "id": "post2",
          },
        ],
      },
    ],
  },
}
`;

exports[`can project user-to-posts relationship: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User]
}

type User {
  id: ID!
  posts: [Post!]!
}

type Post {
  id: ID!
}
"
`;
